[SR-15922] Conflicting messaging around @preconcurrency usage

|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15922      |
   |Radar             | rdar://problem/90348987         |
   |Original Reporter | @adam-fowler      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

Using swift-5.6-DEVELOPMENT-SNAPSHOT-2022-02-11-a compiler


</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 16b7aaa8b476165beb277ae4ddec6a8b

  </details>





**Issue Description:**


Compiling the "sendable" branch of <https://github.com/soto-project/soto-core> produces this remark

``` java
Building for debugging...
/Users/adamfowler/Developer/macOS/lib/soto-project/soto-core/Sources/SotoCore/AsyncAwaitSupport/AWSClient+EndpointDiscovery+async.swift:18:17: remark: '@preconcurrency' attribute on module 'Logging' is unused
@preconcurrency import Logging
~~~~~~~~~~~~~~~~^
/Users/adamfowler/Developer/macOS/lib/soto-project/soto-core/Sources/SotoCore/AsyncAwaitSupport/AWSClient+EndpointDiscovery+async.swift:19:17: remark: '@preconcurrency' attribute on module 'NIOCore' is unused
@preconcurrency import NIOCore
~~~~~~~~~~~~~~~~^
```

But if I do as suggested I get warnings about non-sendable types

``` java
Building for debugging...
/Users/adamfowler/Developer/macOS/lib/soto-project/soto-core/Sources/SotoCore/AsyncAwaitSupport/AWSClient+EndpointDiscovery+async.swift:251:21: warning: capture of 'logger' with non-sendable type 'Logger' in a `@Sendable` closure
                    logger.trace("Request endpoint")
                    ^
/Users/adamfowler/Developer/macOS/lib/soto-project/soto-core/.build/checkouts/swift-log/Sources/Logging/Logging.swift:38:15: note: struct 'Logger' does not conform to the 'Sendable' protocol
public struct Logger {
              ^
/Users/adamfowler/Developer/macOS/lib/soto-project/soto-core/Sources/SotoCore/AsyncAwaitSupport/AWSClient+EndpointDiscovery+async.swift:18:1: remark: add '@preconcurrency' to suppress 'Sendable'-related warnings from module 'Logging'
import Logging
^
@preconcurrency 
/Users/adamfowler/Developer/macOS/lib/soto-project/soto-core/Sources/SotoCore/AsyncAwaitSupport/AWSClient+EndpointDiscovery+async.swift:252:96: warning: capture of 'eventLoop' with non-sendable type 'EventLoop' in a `@Sendable` closure
                    let endpoint = try await endpointDiscovery.getEndpoint(logger: logger, on: eventLoop).get()
                                                                                               ^
/Users/adamfowler/Developer/macOS/lib/soto-project/soto-core/.build/checkouts/swift-nio/Sources/NIOCore/EventLoop.swift:220:17: note: protocol 'EventLoop' does not conform to the 'Sendable' protocol
public protocol EventLoop: EventLoopGroup {
                ^
/Users/adamfowler/Developer/macOS/lib/soto-project/soto-core/Sources/SotoCore/AsyncAwaitSupport/AWSClient+EndpointDiscovery+async.swift:19:1: remark: add '@preconcurrency' to suppress 'Sendable'-related warnings from module 'NIOCore'
import NIOCore
^
@preconcurrency 

```

--- a/lib/Sema/TypeChecker.cpp
+++ b/lib/Sema/TypeChecker.cpp
@@ -279,6 +279,10 @@ static void diagnoseUnnecessaryPreconcurrencyImports(SourceFile &sf) {
   }
 
   ASTContext &ctx = sf.getASTContext();
+
+  if (ctx.TypeCheckerOpts.SkipFunctionBodies != FunctionBodySkipping::None)
+    return;
+
   for (const auto &import : sf.getImports()) {
     if (import.options.contains(ImportFlags::Preconcurrency) &&
         import.importLoc.isValid() &&

--- a/test/Concurrency/predates_concurrency_import_emitmodule.swift
+++ b/test/Concurrency/predates_concurrency_import_emitmodule.swift
@@ -1,0 +1,28 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend -emit-module -emit-module-path %t/StrictModule.swiftmodule -module-name StrictModule -swift-version 6 %S/Inputs/StrictModule.swift
+// RUN: %target-swift-frontend -emit-module -emit-module-path %t/NonStrictModule.swiftmodule -module-name NonStrictModule %S/Inputs/NonStrictModule.swift
+// RUN: %target-swift-frontend -emit-module -emit-module-path %t/OtherActors.swiftmodule -module-name OtherActors %S/Inputs/OtherActors.swift -disable-availability-checking
+
+// RUN: %target-swift-frontend -emit-module -I %t -verify -primary-file %s -emit-module-path %t/predates_concurrency_import_swiftmodule.swiftmodule -experimental-skip-all-function-bodies
+
+// REQUIRES: concurrency
+// REQUIRES: asserts
+
+@preconcurrency import NonStrictModule
+@preconcurrency import StrictModule
+@preconcurrency import OtherActors
+
+// expected-no-warning
+
+func acceptSendable<T: Sendable>(_: T) { }
+
+@available(SwiftStdlib 5.1, *)
+func test(
+  ss: StrictStruct, ns: NonStrictClass, oma: OtherModuleActor,
+  ssc: SomeSendableClass
+) async {
+  acceptSendable(ss)
+  acceptSendable(ns)
+  acceptSendable(oma)
+  acceptSendable(ssc)
+}

